File: src/system/update.ts
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'

const CACHE_FILE = path.join(os.homedir(), '.config', 'anpan', 'update-cache.json')
const CACHE_TTL_MS = 4 * 60 * 60 * 1000 // 4 hours
const REGISTRY_URL = 'https://registry.npmjs.org/anpan-cli/latest'

export type UpdateCheckResult = {
  updateAvailable: boolean
  latestVersion: string
  currentVersion: string
}

type UpdateCache = {
  lastChecked: number
  latestVersion: string
}

export function isNewerVersion(remote: string, current: string): boolean {
  const parse = (v: string) =>
    v
      .replace(/^v/, '')
      .split('.')
      .map(part => parseInt(part, 10) || 0)

  const [rMaj = 0, rMin = 0, rPatch = 0] = parse(remote)
  const [cMaj = 0, cMin = 0, cPatch = 0] = parse(current)

  if (rMaj !== cMaj) return rMaj > cMaj
  if (rMin !== cMin) return rMin > cMin
  return rPatch > cPatch
}

function readCache(): UpdateCache | null {
  try {
    const data = JSON.parse(fs.readFileSync(CACHE_FILE, 'utf8')) as UpdateCache
    if (typeof data.lastChecked === 'number' && typeof data.latestVersion === 'string') {
      return data
    }
  } catch {}
  return null
}

function writeCache(latestVersion: string): void {
  try {
    fs.mkdirSync(path.dirname(CACHE_FILE), {recursive: true})
    fs.writeFileSync(
      CACHE_FILE,
      JSON.stringify({lastChecked: Date.now(), latestVersion}, null, 2),
      'utf8',
    )
  } catch {}
}

export async function checkUpdate(
  currentVersion: string,
  opts?: {force?: boolean; timeoutMs?: number},
): Promise<UpdateCheckResult | null> {
  const timeoutMs = opts?.timeoutMs ?? 1500
  const now = Date.now()

  if (!opts?.force) {
    const cached = readCache()
    if (cached && now - cached.lastChecked < CACHE_TTL_MS) {
      return {
        updateAvailable: isNewerVersion(cached.latestVersion, currentVersion),
        latestVersion: cached.latestVersion,
        currentVersion,
      }
    }
  }

  try {
    const res = await fetch(REGISTRY_URL, {
      signal: AbortSignal.timeout(timeoutMs),
      headers: {Accept: 'application/json'},
    })
    if (!res.ok) return null
    const data = (await res.json()) as {version?: string}
    if (!data.version) return null

    writeCache(data.version)
    return {
      updateAvailable: isNewerVersion(data.version, currentVersion),
      latestVersion: data.version,
      currentVersion,
    }
  } catch {
    // Offline or network timeout: fall back to cached value if present
    const cached = readCache()
    if (cached) {
      return {
        updateAvailable: isNewerVersion(cached.latestVersion, currentVersion),
        latestVersion: cached.latestVersion,
        currentVersion,
      }
    }
    return null
  }
}

File: internal/system/update.go
package system

import (
	"context"
	"encoding/json"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

const (
	cacheTTL        = 4 * time.Hour
	githubLatestURL = "https://api.github.com/repos/KabosuNeko/anpan/releases/latest"
)

type UpdateCheckResult struct {
	UpdateAvailable bool   `json:"updateAvailable"`
	LatestVersion   string `json:"latestVersion"`
	CurrentVersion  string `json:"currentVersion"`
}

type updateCache struct {
	LastChecked   int64  `json:"lastChecked"`
	LatestVersion string `json:"latestVersion"`
}

func cacheFilePath() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".config", "anpan", "update-cache.json")
}

func parseSemver(v string) (int, int, int) {
	clean := strings.TrimPrefix(v, "v")
	parts := strings.Split(clean, ".")
	var nums [3]int
	for i := 0; i < len(parts) && i < 3; i++ {
		n, _ := strconv.Atoi(parts[i])
		nums[i] = n
	}
	return nums[0], nums[1], nums[2]
}

func IsNewerVersion(remote, current string) bool {
	rMaj, rMin, rPatch := parseSemver(remote)
	cMaj, cMin, cPatch := parseSemver(current)

	if rMaj != cMaj {
		return rMaj > cMaj
	}
	if rMin != cMin {
		return rMin > cMin
	}
	return rPatch > cPatch
}

func readCache() *updateCache {
	data, err := os.ReadFile(cacheFilePath())
	if err != nil {
		return nil
	}
	var cache updateCache
	if err := json.Unmarshal(data, &cache); err == nil && cache.LatestVersion != "" {
		return &cache
	}
	return nil
}

func writeCache(latestVersion string) {
	p := cacheFilePath()
	_ = os.MkdirAll(filepath.Dir(p), 0o755)
	c := updateCache{
		LastChecked:   time.Now().UnixMilli(),
		LatestVersion: latestVersion,
	}
	if data, err := json.MarshalIndent(c, "", "  "); err == nil {
		_ = os.WriteFile(p, append(data, '\n'), 0o644)
	}
}

type CheckUpdateOptions struct {
	Force     bool
	TimeoutMs int
}

func CheckUpdate(ctx context.Context, currentVersion string, opts *CheckUpdateOptions) *UpdateCheckResult {
	timeout := 1500 * time.Millisecond
	if opts != nil && opts.TimeoutMs > 0 {
		timeout = time.Duration(opts.TimeoutMs) * time.Millisecond
	}

	force := opts != nil && opts.Force
	now := time.Now().UnixMilli()

	if !force {
		if c := readCache(); c != nil {
			if time.Duration(now-c.LastChecked)*time.Millisecond < cacheTTL {
				return &UpdateCheckResult{
					UpdateAvailable: IsNewerVersion(c.LatestVersion, currentVersion),
					LatestVersion:   c.LatestVersion,
					CurrentVersion:  currentVersion,
				}
			}
		}
	}

	reqCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	req, err := http.NewRequestWithContext(reqCtx, "GET", githubLatestURL, nil)
	if err != nil {
		return nil
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "anpan/"+currentVersion)

	client := &http.Client{Timeout: timeout}
	resp, err := client.Do(req)
	if err != nil {
		if c := readCache(); c != nil {
			return &UpdateCheckResult{
				UpdateAvailable: IsNewerVersion(c.LatestVersion, currentVersion),
				LatestVersion:   c.LatestVersion,
				CurrentVersion:  currentVersion,
			}
		}
		return nil
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil
	}

	var ghRelease struct {
		TagName string `json:"tag_name"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&ghRelease); err != nil || ghRelease.TagName == "" {
		return nil
	}

	latest := strings.TrimPrefix(ghRelease.TagName, "v")
	writeCache(latest)

	return &UpdateCheckResult{
		UpdateAvailable: IsNewerVersion(latest, currentVersion),
		LatestVersion:   latest,
		CurrentVersion:  currentVersion,
	}
}
